### PR TITLE
Add feature switch for order status update

### PR DESCRIPTION
**Pull request by dumega**
# Description
The order state is for Magento to understand and process the order in a defined workflow. Whereas, the order status is for store owners to understand and process the order in a workflow. The merchant may create any number of custom order statuses to manege the exact order flow, but currently the plugin method Bolt\Boltpay\Plugin\OrderPlugin::beforeSetStatus() has logic to ensure that the order in processing/complete/closed status won't be updated back to pending_review and override the default "pending" order status with the "pending_payment", that causes conflict issue. 

To fix this issue, we add a new feature switch M2_DISALLOW_ORDER_STATUS_OVERRIDE to enable/disable the logic.

Fixes: [(link Jira ticket)](https://app.asana.com/0/951157735838091/1201290052673025/f)

#changelog Add feature switch for order status update

# Type of change

- [x] Bug fix (change which fixes an issue)
- [ ] New feature (change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?
Please validate that you have tested your change in at least one of the following areas:

- [ ] Successfully tested locally (or docker image)
- [ ] Successfully tested on a staging or sandbox server
- [ ] Successfully tested on a merchant's staging server

# For PR Reviewer 
- [ ] Reviewed unit tests to make sure we are using real components rather than mocks as much as possible?
- [ ] For any major change (observer, new Bolt feature, core Magento interaction) we must add a feature switch, did you verify this?

# Checklist:

- [ ] My code follows the style guidelines of this project.
- [ ] I have performed a self-review of my own code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] New and existing unit tests pass locally with my changes.
- [ ] I have created or modified unit tests to sufficiently cover my changes.
- [ ] I have added my Jira ticket link and provided a changelog message.

### CODE REVIEW

File: Helper/FeatureSwitch/Definitions.php [hunk 1]
```diff
@@ -456,5 +461,11 @@ class Definitions
             self::DEFAULT_VAL_KEY => false,
             self::ROLLOUT_KEY     => 100
         ],
+        self::M2_DISALLOW_SPECIFIC_ORDER_STATUS_UPDATE => [
+            self::NAME_KEY        => self::M2_DISALLOW_SPECIFIC_ORDER_STATUS_UPDATE,
+            self::VAL_KEY         => true,
+            self::DEFAULT_VAL_KEY => true,
```
Comment (vitaliyreznikov) [inline Helper/FeatureSwitch/Definitions.php:467]: DEFAULT_VAL_KEY should be false

File: Plugin/OrderPlugin.php [hunk 2]
```diff
@@ -113,6 +121,10 @@ public function beforeSetStatus(Order $subject, $status)
             return [Order::STATE_PROCESSING];
         }
         
+        if (!$this->featureSwitches->isDisallowOrderStatusUpdatedBack()) {
```
Comment (vitaliyreznikov) [inline Plugin/OrderPlugin.php:124]: Could you please refactor this to have false as default value? So we can set true to specific merchant when needed

File: Helper/FeatureSwitch/Definitions.php (hunk 1 above)
Comment (dumega) [reply Helper/FeatureSwitch/Definitions.php:467]: @vitaliyreznikov By default we do not allow updating order status from pending to new again. So this value is set to true.

Comment (vitaliyreznikov) [reply Helper/FeatureSwitch/Definitions.php:467]: Please check how other feature switches enable by default look like. It should be 
```
            self::VAL_KEY         => true,
            self::DEFAULT_VAL_KEY => false,
            self::ROLLOUT_KEY     => 100
```

**Review by danac-gs — APPROVED** [review]
The code LGTM. However, this should be turned on very cautiously.

**Comment by dumega** [comment]
@vitaliyreznikov @danac-gs I change the default values of new feature switch to
```
self::M2_DISALLOW_ORDER_STATUS_OVERRIDE => [
            self::NAME_KEY        => self::M2_DISALLOW_ORDER_STATUS_OVERRIDE,
            self::VAL_KEY         => true,
            self::DEFAULT_VAL_KEY => false,
            self::ROLLOUT_KEY     => 0
        ],
```
So it is disabled by default, the plugin method Bolt\Boltpay\Plugin\OrderPlugin::beforeSetStatus() for overriding order status is essential to Bolt checkout. Unless the merchant requests to disable it, our plugin should always process these logic.

File: Plugin/OrderPlugin.php [hunk 3]
```diff
@@ -113,6 +121,10 @@ public function beforeSetStatus(Order $subject, $status)
             return [Order::STATE_PROCESSING];
         }
         
+        if ($this->featureSwitches->isDisallowOrderStatusOverride()) {
```
Comment (vitaliyreznikov) [inline Plugin/OrderPlugin.php:124]: How about others methods in this file? Do I understand correctly that with this feature switch those methods do not change anything?
If so I suggest two options to improve readability:
- check feature switch and early return in each method
or
- add comment / comments

Comment (dumega) [reply Plugin/OrderPlugin.php:130]: Recharge customer with an existing Bolt credit card is a custom behavior of Bolt plugin,  so feature switch M2_DISALLOW_ORDER_STATUS_OVERRIDE does not affect it. 

Then I have added comments for changes.

File: Plugin/OrderPlugin.php [hunk 4]
```diff
@@ -108,11 +116,21 @@ public function beforeSetStatus(Order $subject, $status)
...
+        // Recharge customer with an existing Bolt credit card is a custom behavior of Bolt plugin,
+        // so feature switch M2_DISALLOW_ORDER_STATUS_OVERRIDE does not affect it. 
         if ($subject->getIsRechargedOrder()) {
             // Check and set the recharged order state to processing
             return [Order::STATE_PROCESSING];
         }
         
+        // The order state is for Magento to understand and process the order in a defined workflow.
+        // Whereas, the order status is for store owners to understand and process the order in a workflow.
+        // The merchant may create any number of custom order statuses to manege the exact order flow.
```
Comment (vitaliyreznikov) [inline Plugin/OrderPlugin.php:128]: typo: "to manage"

**Review by vitaliyreznikov — APPROVED** [review]
LGTM, one small comment to address

**Pull request merged**